Sources/MessageLabel.swift

```diff
@@ -280,31 +293,132 @@ open class MessageLabel: UILabel {
+        let checkingTypes = detectorTypes.reduce(0) { $0 | $1.textCheckingType.rawValue }
+        let detector = try? NSDataDetector(types: checkingTypes)
+
+        return detector?.matches(in: text.string, options: [], range: NSRange(location: 0, length: text.length))
+    }
+
+    private func setRangesForDetectors(in checkingResults: [NSTextCheckingResult]) {
+
+        for result in checkingResults {
 
             switch result.resultType {
             case NSTextCheckingResult.CheckingType.address:
-                var ranges = activeDetectorRanges[.address] ?? []
+                var ranges = rangesForDetectors[.address] ?? []
                 ranges.append(result.range)
-                activeDetectorRanges.updateValue(ranges, forKey: .address)
+                rangesForDetectors.updateValue(ranges, forKey: .address)
             case NSTextCheckingResult.CheckingType.date:
-                var ranges = activeDetectorRanges[.date] ?? []
+                var ranges = rangesForDetectors[.date] ?? []
                 ranges.append(result.range)
-                activeDetectorRanges.updateValue(ranges, forKey: .date)
+                rangesForDetectors.updateValue(ranges, forKey: .date)
             case NSTextCheckingResult.CheckingType.phoneNumber:
-                var ranges = activeDetectorRanges[.phoneNumber] ?? []
+                var ranges = rangesForDetectors[.phoneNumber] ?? []
                 ranges.append(result.range)
-                activeDetectorRanges.updateValue(ranges, forKey: .phoneNumber)
+                rangesForDetectors.updateValue(ranges, forKey: .phoneNumber)
             case NSTextCheckingResult.CheckingType.link:
-                var ranges = activeDetectorRanges[.url] ?? []
+                var ranges = rangesForDetectors[.url] ?? []
                 ranges.append(result.range)
-                activeDetectorRanges.updateValue(ranges, forKey: .url)
+                rangesForDetectors.updateValue(ranges, forKey: .url)
             default:
                 fatalError("Received an unrecognized NSTextCheckingResult.CheckingType")
             }
+
+        }
+        
+    }
+
+    // MARK: - Gesture Handling
+
+    private func stringIndex(at location: CGPoint) -> Int? {
+        guard textStorage.length > 0 else { return nil }
+
+        var location = location
+        let textOffset = CGPoint(x: textInsets.left, y: textInsets.right)
+
+        location.x -= textOffset.x
+        location.y -= textOffset.y
+
+        let glyphIndex = layoutManager.glyphIndex(for: location, in: textContainer)
+
+        let lineRect = layoutManager.lineFragmentUsedRect(forGlyphAt: glyphIndex, effectiveRange: nil)
+
+        if lineRect.contains(location) {
+            return layoutManager.characterIndexForGlyph(at: glyphIndex)
+        } else {
+            return nil
+        }
+        
+    }
+
+    private func setupGestureRecognizers() {
+
+        print("Setting up gesture recognizers")
```

Comment:
We probably don't want to leave `print` statements in the final code that will clutter up the command line. These will build up in time and if it is working provide no value.